crates/macros/src/query.rs

```diff
@@ -0,0 +1,123 @@
+use {
+    core::panic,
+    proc_macro::TokenStream,
+    quote::{quote, ToTokens},
+    syn::{parse_macro_input, Data, DeriveInput, Ident, Type},
+};
+
+pub fn process(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let name = input.ident;
+
+    let Data::Enum(data) = input.data else {
+        panic!("only Enums are supported")
+    };
+
+    let path = input
+        .attrs
+        .iter()
+        .find(|attr| attr.path().get_ident().unwrap() == "query_path_override")
+        .map(|attr| {
+            attr.parse_args::<proc_macro2::TokenStream>()
+                .expect("query_path_override must be a valid path")
+        })
+        .unwrap_or(quote! { ::grug });
+
+    let mut generated_structs = Vec::new();
+
+    let mut impl_req_to_enum = Vec::new();
+
+    let mut impl_trait_response = Vec::new();
+
+    let mut impl_as_query_msg = Vec::new();
+
+    for variant in data.variants.into_iter() {
+        let return_type: Type = variant
+            .attrs
+            .iter()
+            .find(|attr| attr.path().get_ident().unwrap() == "returns")
+            .expect("returns attribute missing")
+            .parse_args()
+            .expect("only one type supported");
+
+        let variant_name = &variant.ident;
+
+        let request = match variant.fields {
+            syn::Fields::Named(variant_ty) => {
+                let struct_name =
+                    Ident::new(&format!("{variant_name}Request"), variant.ident.span());
+
+                let mut fields_struct_definition = Vec::new();
+                let mut fields_struct_into = Vec::new();
+
+                for field in variant_ty.named {
+                    let field_name = &field.ident;
+                    let field_type = &field.ty;
+
+                    fields_struct_definition.push(quote! {
+                        pub #field_name: #field_type,
+                    });
+
+                    fields_struct_into.push(quote! {
+                        #field_name: val.#field_name,
+                    });
+                }
+
+                // Generate the struct definition
+                let struct_def = quote! {
+                    pub struct #struct_name {
+                        #(#fields_struct_definition)*
+                    }
+                };
```

Comment:
How about instead of generating a response type for each `QueryMsg` variant, we generate a single `QueryResponse` type?

E.g. for this query message:

```rust
#[derive(grug::Query)]
#[grug::derive(serde)]
enum QueryMsg {
    #[returns(String)]
    Foo {},
    #[returns(u64)]
    Bar {},
}
```

We generate this:

```rust
#[grug::derive(serde)]
enum QueryResponse {
    Foo(String),
    Bar(u64),
}
```

Reply:
Because the compiler is not able to derive the variant of the enum. The purpose of having a bind of Request => Response is not to guess (or match the Response enum variant)

Reply:
If the QueryFooRequest return a String, is better to have the String returned by the query. Having and Enum requires to check if the variant is Foo and then take the String